File: core/src/com/broll/gainea/server/core/map/ExpansionFactory.java
package com.broll.gainea.server.core.map;

import com.broll.gainea.client.ui.render.ExpansionRender;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public abstract class ExpansionFactory {

    private Expansion expansion;
    protected Coordinates baseCoordinates;
    private List<ExpansionFactory> connectedExpansions = new ArrayList<>();
    private Map<AreaID, Area> areas = new HashMap<>();
    private Map<ContinentID, Continent> continents = new HashMap<>();
    private Map<IslandID, Island> islands = new HashMap<>();

    public ExpansionFactory(ExpansionType type) {
        expansion = new Expansion();
        expansion.setType(type);
        setBaseCoordinates(0, 0);
    }

    protected void setBaseCoordinates(float x, float y) {
        baseCoordinates = new Coordinates(x, y);
        expansion.setCoordinates(baseCoordinates);
    }

    protected Coordinates coord(float x, float y) {
        Coordinates c = new Coordinates(x / 100f, y / 100f);
        c.shift(baseCoordinates.getX(), baseCoordinates.getY());
        return c;
    }

    protected void connect(AreaID... areas) {
        for (AreaID a1 : areas) {
            for (AreaID a2 : areas) {
                if (a1 != a2) {
                    Area area1 = get(a1);
                    Area area2 = get(a2);
                    area1.addAdjacentLocation(area2);
                }
            }
        }
    }

    public abstract ExpansionRender createRender();

    protected abstract void init();

    protected abstract void connectWithExpansion(ExpansionFactory expansion);

    public final void connectExpansion(ExpansionFactory expansion) {
        this.connectedExpansions.add(expansion);
        this.connectWithExpansion(expansion);
    }

    protected IslandID island(IslandID id, String name, List<AreaID> areaIds) {
        Island island = new Island(id);
        List<Area> areas = areaIds.stream().map(this::get).collect(Collectors.toList());
        areas.forEach(area -> area.setContainer(island));
        island.setAreas(areas);
        island.setName(name);
        this.islands.put(id, island);
        return id;
    }

    protected ContinentID continent(ContinentID id, String name, List<AreaID> areaIds) {
        Continent continent = new Continent(id);
        List<Area> areas = areaIds.stream().map(this::get).collect(Collectors.toList());
        areas.forEach(area -> area.setContainer(continent));
        continent.setAreas(areas);
        continent.setName(name);
        this.continents.put(id, continent);
        return id;
    }

    protected AreaID area(AreaID id, String name, AreaType type, float x, float y) {
        Area area = new Area(id);
        area.setName(name);
        area.setType(type);
        area.setCoordinates(coord(x, y));
        this.areas.put(id, area);
        return id;
    }

    private Ship ship(Location from, Location to, float x, float y) {
        Ship ship = new Ship();
        ship.setFrom(from);
        ship.setTo(to);
        if (from instanceof Area) {
            ((Area) from).addAdjacentLocation(ship);
        }
        if (to instanceof Area) {
            ((Area) to).addAdjacentLocation(ship);
        }
        ship.setCoordinates(coord(x, y));
        return ship;
    }

    protected Ship ship(AreaID from, AreaID to, float x, float y) {
        Ship ship = ship(get(from), get(to), x, y);
        get(from).getContainer().getShips().add(ship);
        return ship;
    }

    protected List<Ship> ships(AreaID fromId, AreaID toId, float[] x, float[] y) {
        List<Ship> ships = new ArrayList<>();
        Location current = get(fromId);
        Area to = get(toId);
        for (int i = 0; i < x.length; i++) {
            ships.add(ship(current, null, x[i], y[i]));
            current = ships.get(i);
        }
        for (int i = 0; i < x.length; i++) {
            Location next = to;
            if (i < x.length - 1) {
                next = ships.get(i + 1);
            } else {
                to.addAdjacentLocation(ships.get(i));
            }
            ships.get(i).setTo(next);
        }
        get(fromId).getContainer().getShips().addAll(ships);
        return ships;
    }

    protected <T> List<T> list(T... t) {
        return Arrays.asList(t);
    }

    public Expansion create() {
        List<AreaCollection> contents = new ArrayList<>();
        init();
        contents.addAll(islands.values());
        contents.addAll(continents.values());
        expansion.setContents(contents);
        return expansion;
    }

    private Area get(AreaID id) {
        Area area = areas.get(id);
        if (area != null) {
            return area;
        } else {
            for (ExpansionFactory e : connectedExpansions) {
                area = e.areas.get(id);
                if (area != null) {
                    return area;
                }
            }
        }
        return null;
    }

    private Continent get(ContinentID id) {
        Continent area = continents.get(id);
        if (area != null) {
            return area;
        } else {
            for (ExpansionFactory e : connectedExpansions) {
                area = e.continents.get(id);
                if (area != null) {
                    return area;
                }
            }
        }
        return null;
    }

    private Island get(IslandID id) {
        Island area = islands.get(id);
        if (area != null) {
            return area;
        } else {
            for (ExpansionFactory e : connectedExpansions) {
                area = e.islands.get(id);
                if (area != null) {
                    return area;
                }
            }
        }
        return null;
    }
}

File: server/src/main/com/broll/gainea/server/core/map/ExpansionFactory.kt
package com.broll.gainea.server.core.map

abstract class ExpansionFactory(val type: ExpansionType, val texture: String) {
    private val expansion: Expansion
    protected val baseCoordinates: Coordinates = Coordinates(0f, 0f)
    private val connectedExpansions = mutableListOf<ExpansionFactory>()
    private val areas = HashMap<AreaID, Area>()
    private val continents = HashMap<ContinentID, Continent>()
    private val islands = HashMap<IslandID, Island>()

    init {
        expansion = Expansion(type, baseCoordinates)
    }

    private fun coord(x: Float, y: Float): Coordinates {
        val c = Coordinates(x / 100f, y / 100f)
        c.shift(baseCoordinates.rootX, baseCoordinates.rootY)
        return c
    }

    protected fun connect(vararg areas: AreaID) {
        for (a1 in areas) {
            for (a2 in areas) {
                if (a1 !== a2) {
                    val area1 = get(a1)
                    val area2 = get(a2)
                    area1.addAdjacentLocation(area2)
                }
            }
        }
    }

    protected abstract fun init()
    protected abstract fun connectWithExpansion(expansion: ExpansionFactory)
    fun connectExpansion(expansion: ExpansionFactory) {
        connectedExpansions.add(expansion)
        connectWithExpansion(expansion)
    }

    protected fun island(id: IslandID, name: String, areaIds: List<AreaID>): IslandID {
        val island = Island(name, id)
        val areas = areaIds.map { areas[it]!! }
        areas.forEach { it.container = island }
        island.areas += areas
        islands[id] = island
        return id
    }

    protected fun continent(id: ContinentID, name: String, areaIds: List<AreaID>): ContinentID {
        val continent = Continent(name, id)
        val areas = areaIds.map { areas[it]!! }
        areas.forEach { it.container = continent }
        continent.areas += areas
        continents[id] = continent
        return id
    }

    protected fun area(id: AreaID, name: String, type: AreaType, x: Float, y: Float): AreaID {
        val area = Area(id, type, name, coord(x, y))
        areas[id] = area
        return id
    }

    private fun ship(from: Location, to: Location, x: Float, y: Float): Ship {
        val ship = Ship(coord(x, y))
        ship.from = from
        ship.to = to
        if (from is Area) {
            from.addAdjacentLocation(ship)
        }
        if (to is Area) {
            to.addAdjacentLocation(ship)
        }
        return ship
    }

    protected fun ship(from: AreaID, to: AreaID, x: Float, y: Float): Ship {
        val ship = ship(get(from), get(to), x, y)
        get(from).container.ships.add(ship)
        ship.container = get(from).container
        return ship
    }

    protected fun ships(fromId: AreaID, toId: AreaID, x: FloatArray, y: FloatArray): List<Ship> {
        val container = get(fromId).container
        val ships = x.indices.map { i ->
            val ship = Ship(coord(x[i], y[i]))
            ship.container = container
            ship
        }
        val from = get(fromId)
        val to = get(toId)
        ships.forEachIndexed { i, ship ->
            if (i == 0) {
                ship.from = from
            } else {
                ship.from = ships[i - 1]
            }
            if (i == ships.lastIndex) {
                ship.to = to
            } else {
                ship.to = ships[i + 1]
            }
        }
        from.addAdjacentLocation(ships.first())
        to.addAdjacentLocation(ships.last())
        container.ships.addAll(ships)
        return ships
    }

    fun create(): Expansion {
        init()
        expansion.contents += continents.values
        expansion.contents += islands.values
        islands.values.forEach { it.init(expansion) }
        continents.values.forEach { it.init(expansion) }
        return expansion
    }

    private operator fun get(id: AreaID): Area {
        var area = areas[id]
        if (area != null) {
            return area
        } else {
            for (e in connectedExpansions) {
                area = e.areas[id]
                if (area != null) {
                    return area
                }
            }
        }
        throw RuntimeException("unknown id")
    }

    private operator fun get(id: ContinentID): Continent {
        var area = continents[id]
        if (area != null) {
            return area
        } else {
            for (e in connectedExpansions) {
                area = e.continents[id]
                if (area != null) {
                    return area
                }
            }
        }
        throw RuntimeException("unknown id")
    }

    private operator fun get(id: IslandID): Island {
        var area = islands[id]
        if (area != null) {
            return area
        } else {
            for (e in connectedExpansions) {
                area = e.islands[id]
                if (area != null) {
                    return area
                }
            }
        }
        throw RuntimeException("unknown id")
    }
}
